core/src/main/scala/kafka/server/ReplicaAlterDirThread.scala

```diff
@@ -0,0 +1,238 @@
+        .format(topicPartition, futureReplica.logEndOffset.messageOffset, currentReplicaStartOffset))
+      val offsetToFetch = Math.max(currentReplicaStartOffset, futureReplica.logEndOffset.messageOffset)
+      // Only truncate the log when current replica's log start offset is greater than future replica's log end offset.
+      if (currentReplicaStartOffset > futureReplica.logEndOffset.messageOffset)
+        partition.truncateFullyAndStartAt(currentReplicaStartOffset, isFuture = true)
+      offsetToFetch
+    }
+  }
+
+  def handlePartitionsWithErrors(partitions: Iterable[TopicPartition]) {
+    warn("Error while fetching data for future partitions %s".format(partitions.mkString(",")))
+  }
+
+  def buildLeaderEpochRequest(allPartitions: Seq[(TopicPartition, PartitionFetchState)]): Map[TopicPartition, Int] = {
+    val result = allPartitions
+      .filter { case (_, state) => state.isTruncatingLog }
+      .map { case (tp, _) => tp -> replicaMgr.getReplicaOrException(tp, Request.FutureLocalReplicaId).epochs.get.latestEpoch }.toMap
+    result
+  }
+
+  def fetchEpochsFromLeader(partitions: Map[TopicPartition, Int]): Map[TopicPartition, EpochEndOffset] = {
+    partitions.map { case (tp, epoch) =>
+      tp -> new EpochEndOffset(Errors.NONE, replicaMgr.getReplica(tp).get.epochs.get.endOffsetFor(epoch))
+    }
+  }
+
+  def maybeTruncate(fetchedEpochs: Map[TopicPartition, EpochEndOffset]): Map[TopicPartition, Long] = {
+    val truncationPoints = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+
+    fetchedEpochs.foreach { case (topicPartition, epochOffset) =>
+      val currentReplica = replicaMgr.getReplicaOrException(topicPartition)
+      val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+      val partition = replicaMgr.getPartition(topicPartition).get
+
+      if (epochOffset.hasError) {
+        info(s"Retrying leaderEpoch request for partition $topicPartition as the leader reported an error: ${epochOffset.error}")
+      } else {
+        val truncationOffset =
+          if (epochOffset.endOffset() == UNDEFINED_EPOCH_OFFSET)
+            futureReplica.highWatermark.messageOffset
+          else if (epochOffset.endOffset >= futureReplica.logEndOffset.messageOffset)
+            futureReplica.logEndOffset.messageOffset
+          else
+            epochOffset.endOffset
+        partition.truncateTo(truncationOffset, isFuture = true)
+        truncationPoints.put(topicPartition, truncationOffset)
+      }
+    }
+
+    truncationPoints
+  }
+
+  def buildFetchRequest(partitionMap: Seq[(TopicPartition, PartitionFetchState)]): FetchRequest = {
+    // Only include replica in the fetch request if it is not throttled.
+    val iterator = partitionMap.filter { case (topicPartition, partitionFetchState) =>
+      partitionFetchState.isReadyForFetch && !quota.isQuotaExceeded
+    }.sortWith { (left, right) =>
```

Comment:
A related concern here is that we are paying the sort overhead on every fetch request. If there are lots of partitions, this could add non-trivial overhead. The queue approach where the ReplicaAlterDirThread takes one partition at a time from the queue obviates the need for that. It's fine to optimize this in a followup patch though.

Reply:
@junrao Good point. I have updated the patch with the following code. The new implementation should require only one pass of the partitionMap with O(n) time complexity, which is same as the time complexity of ReplicaFetcherThread.buildFetchRequest().

Reply:
Here is the code. The idea is that we only need the maximum partition in the filtered partitionMap.

```
    val maxPartitionOpt = partitionMap.filter { case (topicPartition, partitionFetchState) =>
      partitionFetchState.isReadyForFetch && !quota.isQuotaExceeded
    }.reduceLeftOption { (left, right) =>
      if ((left._1.topic > right._1.topic()) || (left._1.topic == right._1.topic() && left._1.partition() >= right._1.partition()))
        left
      else
        right
    }
```